decoder.go

```diff
@@ -146,60 +179,71 @@ func (d *decoder) decode(nalu []byte) (image.Image, error) {
 	avPacket.size = C.int(len(nalu))
 	res := C.avcodec_send_packet(d.codecCtx, &avPacket)
 	if res < 0 {
+		//nolint:nilnil // avcodec_send_packet raises non-critical errors
 		return nil, nil
 	}
 
 	// receive frame if available
-	res = C.avcodec_receive_frame(d.codecCtx, d.srcFrame)
+	res = C.avcodec_receive_frame(d.codecCtx, d.src.frame)
 	if res < 0 {
+		//nolint:nilnil // avcodec_receive_frame raises non-critical errors
 		return nil, nil
 	}
 
-	// if frame size has changed, allocate needed objects
-	if d.dstFrame == nil || d.dstFrame.width != d.srcFrame.width || d.dstFrame.height != d.srcFrame.height {
-		if d.dstFrame != nil {
-			C.av_frame_free(&d.dstFrame)
-		}
+	// get a frame from the pool
+	d.dst = d.avFramePool.Get().(*avFrameWrapper)
+	if d.dst == nil {
+		return nil, errors.New("failed to obtain AVFrame from pool")
+	}
 
+	// if frame size has changed, reinitialize dstFrame
+	if d.dst.frame.width != d.src.frame.width || d.dst.frame.height != d.src.frame.height {
 		if d.swsCtx != nil {
 			C.sws_freeContext(d.swsCtx)
 		}
 
-		d.dstFrame = C.av_frame_alloc()
-		d.dstFrame.format = C.AV_PIX_FMT_RGBA
-		d.dstFrame.width = d.srcFrame.width
-		d.dstFrame.height = d.srcFrame.height
-		d.dstFrame.color_range = C.AVCOL_RANGE_JPEG
-		res = C.av_frame_get_buffer(d.dstFrame, 1)
+		dstFrame, err := allocateAVFrame()
+		if err != nil {
+			return nil, errors.Errorf("AV frame allocation error while decoding: %v", err)
+		}
+		d.dst = dstFrame
+		d.dst.frame.format = C.AV_PIX_FMT_RGBA
+		d.dst.frame.width = d.src.frame.width
+		d.dst.frame.height = d.src.frame.height
+		d.dst.frame.color_range = C.AVCOL_RANGE_JPEG
+		res = C.av_frame_get_buffer(d.dst.frame, 1)
 		if res < 0 {
 			return nil, errors.New("av_frame_get_buffer() err")
 		}
 
-		d.swsCtx = C.sws_getContext(d.srcFrame.width, d.srcFrame.height, C.AV_PIX_FMT_YUV420P,
-			d.dstFrame.width, d.dstFrame.height, (int32)(d.dstFrame.format), C.SWS_BILINEAR, nil, nil, nil)
+		d.swsCtx = C.sws_getContext(d.src.frame.width, d.src.frame.height, C.AV_PIX_FMT_YUV420P,
+			d.dst.frame.width, d.dst.frame.height, (int32)(d.dst.frame.format), C.SWS_BILINEAR, nil, nil, nil)
 		if d.swsCtx == nil {
 			return nil, errors.New("sws_getContext() err")
 		}
 	}
 
+	dstFrameSize := C.av_image_get_buffer_size((int32)(d.dst.frame.format), d.dst.frame.width, d.dst.frame.height, 1)
+	dstFramePtr := (*[1 << 30]uint8)(unsafe.Pointer(d.dst.frame.data[0]))[:dstFrameSize:dstFrameSize]
+
 	// convert frame from YUV420 to RGB
-	res = C.sws_scale(d.swsCtx, frameData(d.srcFrame), frameLineSize(d.srcFrame),
-		0, d.srcFrame.height, frameData(d.dstFrame), frameLineSize(d.dstFrame))
+	res = C.sws_scale(d.swsCtx, frameData(d.src.frame), frameLineSize(d.src.frame),
+		0, d.src.frame.height, frameData(d.dst.frame), frameLineSize(d.dst.frame))
 	if res < 0 {
```

Comment:
why did we change the order of `av_image_get_buffer_size` and `sws_scale`?

Reply:
This was unintentional. Switching back